pkg/controller/telemetry/metric.go

```diff
@@ -0,0 +1,307 @@
+	trafficLabels.destinationService = dstIP
+
+	trafficLabels.requestProtocol = "tcp"
+	trafficLabels.responseFlags = "-"
+	trafficLabels.connectionSecurityPolicy = "mutual_tls"
+
+	return trafficLabels, nil
+}
+
+func (m *MetricController) getWorkloadByAddress(address []byte) (*workloadapi.Workload, string) {
+	networkAddr := cache.NetworkAddress{}
+	networkAddr.Address, _ = netip.AddrFromSlice(address)
+	workload := m.workloadCache.GetWorkloadByAddr(networkAddr)
+	if workload == nil {
+		log.Warnf("get worload from ip %v FAILED", address)
+		return nil, ""
+	}
+	return workload, networkAddr.Address.String()
+}
+
+func buildMetricFromWorkload(dstWorkload, srcWorkload *workloadapi.Workload) commonTrafficLabels {
+	if dstWorkload == nil || srcWorkload == nil {
+		return commonTrafficLabels{}
+	}
+
+	trafficLabels := commonTrafficLabels{}
+
+	trafficLabels.destinationServiceNamespace = dstWorkload.Namespace
+	trafficLabels.destinationServiceName = dstWorkload.Name
+	trafficLabels.destinationWorkload = dstWorkload.WorkloadName
+	trafficLabels.destinationCanonicalService = dstWorkload.CanonicalName
+	trafficLabels.destinationCanonicalRevision = dstWorkload.CanonicalRevision
+	trafficLabels.destinationWorkloadNamespace = dstWorkload.Namespace
+	trafficLabels.destinationApp = dstWorkload.CanonicalName
+	trafficLabels.destinationVersion = dstWorkload.CanonicalRevision
+	trafficLabels.destinationCluster = dstWorkload.ClusterId
+
+	trafficLabels.sourceWorkload = srcWorkload.WorkloadName
+	trafficLabels.sourceCanonicalService = srcWorkload.CanonicalName
+	trafficLabels.sourceCanonicalRevision = srcWorkload.CanonicalRevision
+	trafficLabels.sourceWorkloadNamespace = srcWorkload.Namespace
+	trafficLabels.sourceApp = srcWorkload.CanonicalName
+	trafficLabels.sourceVersion = srcWorkload.CanonicalRevision
+	trafficLabels.sourceCluster = srcWorkload.ClusterId
+
+	trafficLabels.destinationPrincipal = buildPrincipal(dstWorkload)
+	trafficLabels.sourcePrincipal = buildPrincipal(srcWorkload)
+
+	return trafficLabels
+}
+
+func buildPrincipal(workload *workloadapi.Workload) string {
+	if workload.TrustDomain != "" && workload.ServiceAccount != "" && workload.Namespace != "" {
+		return fmt.Sprintf("spiffe://%s/ns/%s/sa/%s", workload.TrustDomain, workload.Namespace, workload.ServiceAccount)
+	}
+	return "-"
+}
+
+func buildMetricsToPrometheus(data requestMetric, labels commonTrafficLabels) {
+	connectionOpened, connectionClosed, receivedBytes, sentBytes := []byte{}, []byte{}, []byte{}, []byte{}
+	connectionOpened = binary.LittleEndian.AppendUint32(connectionOpened, data.connectionOpened)
+	connectionClosed = binary.LittleEndian.AppendUint32(connectionClosed, data.connectionClosed)
+	receivedBytes = binary.LittleEndian.AppendUint32(receivedBytes, data.receivedBytes)
+	sentBytes = binary.LittleEndian.AppendUint32(sentBytes, data.sentBytes)
+
+	commonLabels := commonTrafficLabels2map(&labels)
+	tcpConnectionOpened.With(commonLabels).Set(byteToFloat64(connectionOpened))
```

Comment:
why canot you set with float64(data.connectionOpened), and for other values it is similar